homeassistant/components/telegram_bot/__init__.py

```diff
@@ -320,15 +321,49 @@ async def async_send_telegram_message(service: ServiceCall) -> ServiceResponse:
                 service.hass.config_entries.async_entries(DOMAIN)
             )
 
-            if len(config_entries) == 1:
-                config_entry = config_entries[0]
-
-            if len(config_entries) > 1:
+            if len(config_entries) == 0:
                 raise ServiceValidationError(
-                    "Multiple config entries found. Please specify the Telegram bot to use.",
+                    "No Telegram Bot config entries found. Please set up a Telegram Bot first.",
                     translation_domain=DOMAIN,
-                    translation_key="multiple_config_entry",
+                    translation_key="no_config_entries",
                 )
+            if len(config_entries) == 1:
+                config_entry = config_entries[0]
+            else:
+                # Multiple config entries: try to find matching bot
+                target = kwargs.get(ATTR_TARGET)
+                if target is not None:
+                    # Convert single target to list for consistent handling
+                    if isinstance(target, int):
+                        target = [target]
+
+                    # Find config entries that contain any of the target chat_ids
+                    matching_entries = []
+                    for entry in config_entries:
+                        if hasattr(entry, "runtime_data") and entry.runtime_data:
+                            entry_chat_ids = [
```

Comment:
Here as well, we should use a set. Member checking between sets is much faster than between lists.